pkg/groupmapper/groupmapper.go

```diff
@@ -0,0 +1,207 @@
+
+// UserInfoGroupsWrapper wraps a UserInfo object in order to add extra groups
+// retrieved from the identity providers
+type UserInfoGroupsWrapper struct {
+	userInfo         kuser.Info
+	additionalGroups sets.String
+}
+
+func (w *UserInfoGroupsWrapper) GetName() string {
+	return w.userInfo.GetName()
+}
+
+func (w *UserInfoGroupsWrapper) GetUID() string {
+	return w.userInfo.GetUID()
+}
+
+func (w *UserInfoGroupsWrapper) GetExtra() map[string][]string {
+	return w.userInfo.GetExtra()
+}
+
+func (w *UserInfoGroupsWrapper) GetGroups() []string {
+	groups := w.additionalGroups.Union(sets.NewString(w.userInfo.GetGroups()...))
+	return groups.List()
+}
+
+// UserGroupsMapper wraps a UserIdentityMapper with a struct that's capable to
+// create the groups for a given user based on the provided UserIdentityInfo
+type UserGroupsMapper struct {
+	delegatedUserMapper authapi.UserIdentityMapper
+	groupsClient        userclient.GroupInterface
+	groupsCache         *usercache.GroupCache
+	groupsSynced        func() bool
+}
+
+func NewUserGroupsMapper(delegate authapi.UserIdentityMapper, groupsClient userclient.GroupInterface, groupInformer userinformer.GroupInformer) *UserGroupsMapper {
+	return &UserGroupsMapper{
+		delegatedUserMapper: delegate,
+		groupsClient:        groupsClient,
+		groupsCache:         usercache.NewGroupCache(groupInformer),
+		groupsSynced:        groupInformer.Informer().HasSynced,
+	}
+}
+
+func (m *UserGroupsMapper) UserFor(identityInfo authapi.UserIdentityInfo) (kuser.Info, error) {
+	userInfo, err := m.delegatedUserMapper.UserFor(identityInfo)
+	if err != nil {
+		return userInfo, err
+	}
+
+	identityGroups := sets.NewString(identityInfo.GetProviderGroups()...)
+	if err := m.processGroups(identityInfo.GetProviderName(), identityInfo.GetProviderPreferredUserName(), identityGroups); err != nil {
+		return nil, err
+	}
+
+	return &UserInfoGroupsWrapper{
+		userInfo:         userInfo,
+		additionalGroups: identityGroups,
+	}, nil
+}
+
+func (m *UserGroupsMapper) processGroups(idpName, username string, groups sets.String) error {
+	err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+		return m.groupsSynced(), nil
+	})
+	if err != nil {
+		return err
+	}
+
+	cachedGroups, err := m.groupsCache.GroupsFor(username)
+	if err != nil {
+		return err
+	}
+
+	removeGroups, addGroups := groupsDiff(cachedGroups, groups)
+	for _, g := range removeGroups {
+		if err := m.removeUserFromGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	for _, g := range addGroups {
+		if err := m.addUserToGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// FIXME: work better with annotations
+// FIXME: add retries
+// FIXME: use idpName
+func (m *UserGroupsMapper) removeUserFromGroupWithRetries(idpName, username, group string) error {
+	updatedGroup, err := m.groupsClient.Get(context.TODO(), group, metav1.GetOptions{})
```

Comment:
Shouldn't we bubble up the ctx?

Reply:
the context is unfortunately not wired through the oauth-server, and I fear that to really do it properly we would have to also implement https://github.com/openshift/osin/issues/82